### ISSUE TITLE
OS X instructions

### ISSUE THREAD

**Issue by ryansuchocki** [ryansuchocki/microscheme#5]
I know that people have successfully compiled microscheme on OS X. It would be great if someone with a Mac could comment on the availability of the required tools (avr-gcc, avrdude), test whether microscheme invokes them properly, and generally provide some OS X specific instructions.

**Comment by BorisKourt** [comment]
No more warnings on `make build` of the new version.

Still doesn't upload with `-au` or `-auc`  (sudo or no sudo) 

```sh
boris% sudo ./microscheme -m LEO -d /dev/tty.usbmodem1421 -au examples/BLINK.ms
Microscheme 0.8, (C) Ryan Suchocki
>> Treeshaker: After 4 rounds: 83 globals purged! 22 bytes will be reserved.
>> 18 lines compiled OK
>> Assembling...
>> Uploading...
```
*The above output all appears near instantly* 
There are:
```
BLINK.elf
BLINK.hex
BLINK.ms
BLINK.s
```
In the examples folder, looks like .elf .s .hex were generated. Anything else that I can send you, or anything else I can test with?

(Running OSX 10.10)

**Comment by ryansuchocki** [comment]
Run:
```
avrdude -p -V atmega328p -c arduino -P /dev/tty.usbmodem1421 -b 115200 -D -U flash:w:examples/BLINK.hex:i
```
In the same terminal, and paste the whole output?

**Comment by BorisKourt** [comment]
Hmm getting an odd message:
```sh
avrdude: no programmer has been specified on the command line or the config file
         Specify a programmer using the -c option and try again
```
even though the -c option is specified?

**Comment by ryansuchocki** [comment]
Output of ```avrdude -c ?```?

**Comment by BorisKourt** [comment]
```sh
boriss-air% avrdude -c ?
zsh: no matches found: ?
```
tried without the `?` just in case typo and it just gives me the main `avrdude` list of commands.

`avrdude -v`:
```sh
boris% avrdude -v

avrdude: Version 6.1, compiled on Jan  6 2015 at 17:21:39
         Copyright (c) 2000-2005 Brian Dean, http://www.bdmicro.com/
         Copyright (c) 2007-2014 Joerg Wunsch

         System wide configuration file is "/usr/local/Cellar/avrdude/6.1/etc/avrdude.conf"
         User configuration file is "/Users/boriskourt/.avrduderc"
         User configuration file does not exist or is not a regular file, skipping


avrdude: no programmer has been specified on the command line or the config file
         Specify a programmer using the -c option and try again
```

**Comment by ryansuchocki** [comment]
```avrdude -c ?``` should list the available programmers, but maybe zsh does something special with ```?```. 

Anyway, there's an error in the command. (Sorry!) It should be:

```avrdude -V -p atmega328p -c arduino -P /dev/tty.usbmodem1421 -b 115200 -D -U flash:w:examples/BLINK.hex:i```

**Comment by BorisKourt** [comment]
zsh interpreted `?` as its own command (wasn't aware of this)

Here is `avrdude -c \?` escaped:

```sh
Valid programmers are:
  2232HIO          = FT2232H based generic programmer
  4232h            = FT4232H based generic programmer
  arduino          = Arduino
  arduino-ft232r   = Arduino: FT232R connected to ISP
  atmelice         = Atmel-ICE (ARM/AVR) in JTAG mode
  atmelice_dw      = Atmel-ICE (ARM/AVR) in debugWIRE mode
  atmelice_isp     = Atmel-ICE (ARM/AVR) in ISP mode
  atmelice_pdi     = Atmel-ICE (ARM/AVR) in PDI mode
  avr109           = Atmel AppNote AVR109 Boot Loader
  avr910           = Atmel Low Cost Serial Programmer
  avr911           = Atmel AppNote AVR911 AVROSP
  avrftdi          = FT2232D based generic programmer
  avrisp           = Atmel AVR ISP
  avrisp2          = Atmel AVR ISP mkII
  avrispmkII       = Atmel AVR ISP mkII
  avrispv2         = Atmel AVR ISP V2
  buspirate        = The Bus Pirate
  buspirate_bb     = The Bus Pirate (bitbang interface, supports TPI)
  butterfly        = Atmel Butterfly Development Board
  butterfly_mk     = Mikrokopter.de Butterfly
  bwmega           = BitWizard ftdi_atmega builtin programmer
  C232HM           = FT232H based module from FTDI and Glyn.com.au
  c2n232i          = serial port banging, reset=dtr sck=!rts mosi=!txd miso=!cts
  dasa             = serial port banging, reset=rts sck=dtr mosi=txd miso=cts
  dasa3            = serial port banging, reset=!dtr sck=rts mosi=txd miso=cts
  diecimila        = alias for arduino-ft232r
  dragon_dw        = Atmel AVR Dragon in debugWire mode
  dragon_hvsp      = Atmel AVR Dragon in HVSP mode
  dragon_isp       = Atmel AVR Dragon in ISP mode
  dragon_jtag      = Atmel AVR Dragon in JTAG mode
  dragon_pdi       = Atmel AVR Dragon in PDI mode
  dragon_pp        = Atmel AVR Dragon in PP mode
  flip1            = FLIP USB DFU protocol version 1 (doc7618)
  flip2            = FLIP USB DFU protocol version 2 (AVR4023)
  ft232r           = FT232R Synchronous BitBang
  ft245r           = FT245R Synchronous BitBang
  jtag1            = Atmel JTAG ICE (mkI)
  jtag1slow        = Atmel JTAG ICE (mkI)
  jtag2            = Atmel JTAG ICE mkII
  jtag2avr32       = Atmel JTAG ICE mkII im AVR32 mode
  jtag2dw          = Atmel JTAG ICE mkII in debugWire mode
  jtag2fast        = Atmel JTAG ICE mkII
  jtag2isp         = Atmel JTAG ICE mkII in ISP mode
  jtag2pdi         = Atmel JTAG ICE mkII PDI mode
  jtag2slow        = Atmel JTAG ICE mkII
  jtag3            = Atmel AVR JTAGICE3 in JTAG mode
  jtag3dw          = Atmel AVR JTAGICE3 in debugWIRE mode
  jtag3isp         = Atmel AVR JTAGICE3 in ISP mode
  jtag3pdi         = Atmel AVR JTAGICE3 in PDI mode
  jtagkey          = Amontec JTAGKey, JTAGKey-Tiny and JTAGKey2
  jtagmkI          = Atmel JTAG ICE (mkI)
  jtagmkII         = Atmel JTAG ICE mkII
  jtagmkII_avr32   = Atmel JTAG ICE mkII im AVR32 mode
  lm3s811          = Luminary Micro LM3S811 Eval Board (Rev. A)
  mib510           = Crossbow MIB510 programming board
  mkbutterfly      = Mikrokopter.de Butterfly
  nibobee          = NIBObee
  o-link           = O-Link, OpenJTAG from www.100ask.net
  openmoko         = Openmoko debug board (v3)
  pavr             = Jason Kyle's pAVR Serial Programmer
  pickit2          = MicroChip's PICkit2 Programmer
  ponyser          = design ponyprog serial, reset=!txd sck=rts mosi=dtr miso=cts
  siprog           = Lancos SI-Prog <http://www.lancos.com/siprogsch.html>
  stk500           = Atmel STK500
  stk500hvsp       = Atmel STK500 V2 in high-voltage serial programming mode
  stk500pp         = Atmel STK500 V2 in parallel programming mode
  stk500v1         = Atmel STK500 Version 1.x firmware
  stk500v2         = Atmel STK500 Version 2.x firmware
  stk600           = Atmel STK600
  stk600hvsp       = Atmel STK600 in high-voltage serial programming mode
  stk600pp         = Atmel STK600 in parallel programming mode
  UM232H           = FT232H based module from FTDI and Glyn.com.au
  usbasp           = USBasp, http://www.fischl.de/usbasp/
  usbasp-clone     = Any usbasp clone with correct VID/PID
  usbtiny          = USBtiny simple USB programmer, http://www.ladyada.net/make/usbtinyisp/
  wiring           = Wiring
  xplainedpro      = Atmel AVR XplainedPro in JTAG mode
```

**Comment by BorisKourt** [comment]
```sh
boris% avrdude -V -p atmega328p -c arduino -P /dev/tty.usbmodem1421 -b 115200 -D -U flash:w:examples/BLINK.hex:i
avrdude: stk500_recv(): programmer is not responding
avrdude: stk500_getsync() attempt 1 of 10: not in sync: resp=0x3c
avrdude: stk500_recv(): programmer is not responding
avrdude: stk500_getsync() attempt 2 of 10: not in sync: resp=0x3c
...
```

Did a test with standard Arduino, its responsive. But I have seen this error before.

**Comment by ryansuchocki** [comment]
Which model are you using now?

**Comment by BorisKourt** [comment]
Leonardo (R3 I think)

**Comment by ryansuchocki** [comment]
Could you try with an UNO?

**Comment by BorisKourt** [comment]
```sh
boris% avrdude -V -p atmega328p -c arduino -P /dev/tty.usbmodem1421 -b 115200 -D -U flash:w:examples/BLINK.hex:i

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00s

avrdude: Device signature = 0x1e950f
avrdude: reading input file "examples/BLINK.hex"
avrdude: writing flash (2338 bytes):

Writing | ################################################## | 100% 0.40s

avrdude: 2338 bytes of flash written

avrdude: safemode: Fuses OK (E:00, H:00, L:00)

avrdude done.  Thank you.
```

**Comment by BorisKourt** [comment]
Is this meant to blink on pin 13?

**Comment by ryansuchocki** [comment]
Yes. Does it?

**Comment by BorisKourt** [comment]
No sorry should have said it didn't. Tested with standard Arduino blink and the LED does work on this UNO.

**Comment by ryansuchocki** [comment]
I think that's trying to upload the version assembled for the LEO, to the UNO.

Try:
```./microscheme -m UNO -a examples/BLINK.ms```

and then:
```avrdude -V -p atmega328p -c arduino -P /dev/tty.usbmodem1421 -b 115200 -D -U flash:w:examples/BLINK.hex:i```

**Comment by BorisKourt** [comment]
Ah sorry about that, that worked!

**Comment by ryansuchocki** [comment]
Excellent. Does it all work together?

```./microscheme -m UNO -d /dev/tty.usbmodem1421 -au examples/BLINK.ms```

**Comment by BorisKourt** [comment]
Sadly no luck, same issue as the first time:
```sh
Microscheme 0.8, (C) Ryan Suchocki
>> Treeshaker: After 4 rounds: 83 globals purged! 22 bytes will be reserved.
>> 18 lines compiled OK
>> Assembling...
>> Uploading...
```

**Comment by ryansuchocki** [comment]
Pull and try now?

**Comment by BorisKourt** [comment]
That worked :)

```sh
Microscheme 0.8, (C) Ryan Suchocki
>> Treeshaker: After 4 rounds: 83 globals purged! 22 bytes will be reserved.
>> 18 lines compiled OK
>> Assembling...
>> Uploading...

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.00s

avrdude: Device signature = 0x1e950f
avrdude: reading input file "examples/BLINK.hex"
avrdude: writing flash (2110 bytes):

Writing | ################################################## | 100% 0.35s

avrdude: 2110 bytes of flash written

avrdude: safemode: Fuses OK (E:00, H:00, L:00)

avrdude done.  Thank you.

>> Finished.
```

**Comment by ryansuchocki** [comment]
Fantastic.

The problem was an internal buffer set too small. It never came up before because "/dev/tty.usbmodem1421" is longer than any linux device path!

Have fun! :)

**Comment by BorisKourt** [comment]
Yeah nice! Should I add this to the README or make a wiki page? (With the what the prereqs were)

Also any idea on what to do about the Leonardo? Still experiences the same timeouts even when compiled and set to `LEO`

**Comment by BorisKourt** [comment]
Alright thank you. 

I have many different types of Arduino on hand (both *official* and *not official*), what is proper way to add support for them and test it?

**Comment by ryansuchocki** [comment]
I'll start a new issue.
